File: src/minigames/apocalypse_survival/java/work/lclpnet/ap2/game/apocalypse_survival/util/TargetManager.java
package work.lclpnet.ap2.game.apocalypse_survival.util;

import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.entity.Mob;
import net.minecraft.world.entity.monster.*;
import work.lclpnet.ap2.api.base.Participants;
import work.lclpnet.lobby.game.map.GameMap;

import java.util.Random;

public class TargetManager {

    private final PursuitClass<?>[] pursuits;
    private final PursuitClass<Zombie> zombiePursuit;
    private final PursuitClass<AbstractSkeleton> skeletonPursuit;
    private final PursuitClass<Phantom> phantomPursuit;
    private final PursuitClass<Vindicator> vindicatorPursuit;
    private final MobDensityManager densityManager;

    public TargetManager(Participants participants, GameMap map, Random random) {
        zombiePursuit = new PursuitClass<>(participants, 20);
        skeletonPursuit = new PursuitClass<>(participants, 10);
        phantomPursuit = new PursuitClass<>(participants, 3);
        vindicatorPursuit = new PursuitClass<>(participants, 5);

        pursuits = new PursuitClass[] {zombiePursuit, skeletonPursuit, phantomPursuit, vindicatorPursuit};

        densityManager = new MobDensityManager(map, random);
    }

    public void addZombie(Zombie zombie) {
        this.addMob(zombie);
        zombiePursuit.addMob(zombie);
    }

    public void addSkeleton(AbstractSkeleton skeleton) {
        this.addMob(skeleton);
        skeletonPursuit.addMob(skeleton);
    }

    public void addPhantom(Phantom phantom) {
        this.addMob(phantom);
        phantomPursuit.addMob(phantom);
    }

    public void addVindicator(Vindicator vindicator) {
        this.addMob(vindicator);
        vindicatorPursuit.addMob(vindicator);
    }

    private void addMob(Mob mob) {
        densityManager.startTracking(mob);
    }

    public void removeMob(Mob mob) {
        densityManager.stopTracking(mob);

        switch (mob) {
            case Zombie zombie -> zombiePursuit.removeMob(zombie);
            case Skeleton skeleton -> skeletonPursuit.removeMob(skeleton);
            case Phantom phantom -> phantomPursuit.removeMob(phantom);
            case Vindicator vindicator -> vindicatorPursuit.removeMob(vindicator);
            default -> {}
        }
    }

    public void removeParticipant(ServerPlayer player) {
        for (var pursuit : pursuits) {
            pursuit.removeParticipant(player);
        }
    }

    public void update() {
        densityManager.update();

        for (var pursuit : pursuits) {
            pursuit.update();
        }
    }

    public MobDensityManager getDensityManager() {
        return densityManager;
    }
}

File: src/minigames/apocalypse_survival/src/main/kotlin/work/lclpnet/ap2/game/apocalypse_survival/util/TargetManager.kt
package work.lclpnet.ap2.game.apocalypse_survival.util

import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.entity.Mob
import net.minecraft.world.entity.monster.Phantom
import net.minecraft.world.entity.monster.illager.Vindicator
import net.minecraft.world.entity.monster.skeleton.AbstractSkeleton
import net.minecraft.world.entity.monster.skeleton.Skeleton
import net.minecraft.world.entity.monster.zombie.Zombie
import work.lclpnet.ap2.game.player.Participants
import work.lclpnet.game.map.GameMap
import java.util.*

class TargetManager(participants: Participants, map: GameMap, random: Random) {

    private val zombiePursuit = PursuitClass<Zombie>(participants, 20)
    private val skeletonPursuit = PursuitClass<AbstractSkeleton>(participants, 10)
    private val phantomPursuit = PursuitClass<Phantom>(participants, 3)
    private val vindicatorPursuit = PursuitClass<Vindicator>(participants, 5)
    private val pursuits = arrayOf(zombiePursuit, skeletonPursuit, phantomPursuit, vindicatorPursuit)
    val densityManager = MobDensityManager(map, random)

    fun addZombie(zombie: Zombie) {
        densityManager.startTracking(zombie)
        zombiePursuit.addMob(zombie)
    }

    fun addSkeleton(skeleton: AbstractSkeleton) {
        densityManager.startTracking(skeleton)
        skeletonPursuit.addMob(skeleton)
    }

    fun addPhantom(phantom: Phantom) {
        densityManager.startTracking(phantom)
        phantomPursuit.addMob(phantom)
    }

    fun addVindicator(vindicator: Vindicator) {
        densityManager.startTracking(vindicator)
        vindicatorPursuit.addMob(vindicator)
    }

    fun removeMob(mob: Mob) {
        densityManager.stopTracking(mob)

        when (mob) {
            is Zombie -> zombiePursuit.removeMob(mob)
            is Skeleton -> skeletonPursuit.removeMob(mob)
            is Phantom -> phantomPursuit.removeMob(mob)
            is Vindicator -> vindicatorPursuit.removeMob(mob)
        }
    }

    fun removeParticipant(player: ServerPlayer) {
        for (pursuit in pursuits) {
            pursuit.removeParticipant(player)
        }
    }

    fun update() {
        densityManager.update()

        for (pursuit in pursuits) {
            pursuit.update()
        }
    }
}
